neural-network/PulsedRADAR/mim_wg_pid_nn_dl.py

```diff
@@ -0,0 +1,119 @@
+from moku.instruments import WaveformGenerator, PIDController, NeuralNetwork, Datalogger
+
+# Import libraries for matplotlib in order to plot results
+import matplotlib.pyplot as plt
+
+import time
+
+# Establish connection to Moku:Pro
+m = MultiInstrument('192.168.1.226', force_connect=True, platform_id=4)
+
+# Establish connection to Moku:Go
+mg = WaveformGenerator('192.168.1.47', force_connect=True)
+
+try:
+	cnt = 1
+	# Configure Moku:Go to generate pulsed signal
+	# Pulse parameters are:
+		# Pulse Repetition Interval (PRI) - 10 Hz
+		# Carrier Frequency - 2kHz
+		# Will start with an output of 1Vpp and gradually decrease
+	mg.generate_waveform(channel=1, type='Sine', amplitude=cnt*.1, frequency=2e3)
+	mg.generate_waveform(channel=2, type='Pulse', pulse_width=.005, edge_time=16e-9, amplitude=1, frequency=10)
+	mg.set_modulation(channel=1, type='Amplitude', source='Output2', depth=200)
+
+	# Configure Moku:Pro with Neural Network in MiM
+	wg = m.set_instrument(1, WaveformGenerator)
+	pid = m.set_instrument(2, PIDController)
+	nn = m.set_instrument(3, NeuralNetwork)
+	dl = m.set_instrument(4, Datalogger)
+	connections = [dict(source="Slot1OutB", destination="Slot2InB"),
+				   dict(source="Input1", destination="Slot2InA"),
+	               dict(source="Slot2OutA", destination="Slot3InA"),
+	               dict(source="Slot2OutA", destination="Slot3InB"),
+	               dict(source="Slot2OutA", destination="Slot3InC"),
+	               dict(source="Slot2OutA", destination="Slot3InD"),
+	               dict(source="Slot3OutA", destination="Slot4InD"),
+	               dict(source="Input1", destination="Slot4InB"),
+	               dict(source="Slot2OutA", destination="Slot4InC")
+	               ]
+	m.set_connections(connections=connections)
+
+	# Configure Waveform Generator
+	wg.generate_waveform(channel=2, type='Noise')
+
+	# Configure PID Controller to combine noise with received pulsed waveform
+
+	# # Configures the control matrix:
+	# Channel 1: input 1 gain = 1 dB, input 2 gain = 20 dB
+	pid.set_control_matrix(channel=1, input_gain1=1, input_gain2=20)
+
+	# Configure PID Control loop 1 using frequency response characteristics
+    #   P = 0dB
+	pid.set_by_frequency(channel=1, prop_gain=0)
+	pid.enable_output(1, signal=True, output=True)
+	print(pid.summary())
+
+	# Load network into Neural Network and set inputs and outputs
+	nn.set_input(strict=False, channel=1, low_level=-1, high_level=1)
+	nn.set_input_sample_rate(sample_rate=305000)
+	nn.upload_network("./autoencoder_32.linn")
+	nn.set_output(strict=False, channel=1, enabled=True, low_level=-1, high_level=1)
+	print(nn.summary())
+
+	# Only log data from inputs 2,3,4 per the configuration above
+	dl.enable_input(1, enable=False)
+
+	# Set the sample rate to 500 KSa/s
+	dl.set_samplerate(500e3)
+
+	# Set the acquisition mode
+	dl.set_acquisition_mode(mode='Normal')
+
+	# Gradually increase input pulse from .1Vpp to a max of 2Vpp 
+	while(cnt <= 20):
```

Comment:
`cnt` isn't used for anything other than making this look like a `for` loop, i.e. not used except as a simple counter inside the loop? Prefer this loop be
```
for cnt in range(20):
```
(Same paradigm elsewhere)

Reply:
Old bad habits die hard.  I've never liked the syntax for "for loops" in Python.  I think I have them good now.